src/modules/performance_collector/performance_collector.py

```diff
@@ -0,0 +1,172 @@
+        cc_spec = self.w3.cc.get_config_spec()
+        genesis = self.w3.cc.get_genesis()
+        chain_cfg = ChainConfig(
+            slots_per_epoch=cc_spec.SLOTS_PER_EPOCH,
+            seconds_per_slot=cc_spec.SECONDS_PER_SLOT,
+            genesis_time=genesis.genesis_time,
+        )
+        return ChainConverter(chain_cfg)
+
+    def execute_module(self, last_finalized_blockstamp: BlockStamp) -> ModuleExecuteDelay:
+        converter = self._build_converter()
+
+        # TODO: return comment about finalized_epoch
+        finalized_epoch = EpochNumber(converter.get_epoch_by_slot(last_finalized_blockstamp.slot_number) - 1)
+
+        epochs_range_demand = self.define_epochs_to_process_range(finalized_epoch)
+        if not epochs_range_demand:
+            return ModuleExecuteDelay.NEXT_SLOT
+        start_epoch, end_epoch = epochs_range_demand
+
+        checkpoints = FrameCheckpointsIterator(
+            converter,
+            start_epoch,
+            end_epoch,
+            finalized_epoch,
+        )
+        processor = FrameCheckpointProcessor(self.w3.cc, self.db, converter, last_finalized_blockstamp)
+
+        checkpoint_count = 0
+        for checkpoint in checkpoints:
+            processed_epochs = processor.exec(checkpoint)
+            checkpoint_count += 1
+            logger.info({
+                'msg': 'Checkpoint processing completed',
+                'checkpoint_slot': checkpoint.slot,
+                'processed_epochs': processed_epochs
+            })
+            # Reset BaseOracle cycle timeout to avoid timeout errors during long checkpoints processing
+            self._reset_cycle_timeout()
+
+            if self.new_epochs_range_demand_appeared():
+                logger.info({"msg": "New epochs demand is found during processing"})
+                return ModuleExecuteDelay.NEXT_SLOT
+
+        logger.info({
+            'msg': 'All checkpoints processing completed',
+            'total_checkpoints_processed': checkpoint_count
+        })
+
+        return ModuleExecuteDelay.NEXT_SLOT
+
+    def define_epochs_to_process_range(self, finalized_epoch: EpochNumber) -> tuple[EpochNumber, EpochNumber] | None:
+        max_available_epoch_to_check = finalized_epoch - FrameCheckpointsIterator.CHECKPOINT_SLOT_DELAY_EPOCHS
+        if max_available_epoch_to_check < 0:
+            logger.info({"msg": "No available epochs to process yet"})
+            return None
+
+        start_epoch = EpochNumber(max_available_epoch_to_check)
+        end_epoch = EpochNumber(max_available_epoch_to_check)
+
+        min_epoch_in_db = self.db.min_epoch()
+        max_epoch_in_db = self.db.max_epoch()
+        if not min_epoch_in_db and not max_epoch_in_db:
+            logger.info({
+                "msg": "Empty Performance Collector DB. Start with the default range calculation",
+                "start_epoch": start_epoch,
+                "end_epoch": end_epoch
+            })
+            return start_epoch, end_epoch
```

Comment:
A request for an epoch range might already exist though.

Reply:
It was implemented intentionally, but might be revised.